.github/CONTRIBUTING.md

```diff
@@ -1,6 +1,130 @@
+
+Example of a good pull request: [PR#2332](https://github.com/Rdatatable/data.table/pull/2332). It has a NEWS entry. It passed existing tests and added a new one. One test was removed but the PR description clearly explained why upfront (without us having to ask). Benchmark results were included, which made the need for the change compelling. We didn't need to run anything ourselves. Everything was including in one PR in one place. In short, it was a pleasure to review and merge.
+
+### Coding Style
+
+A few minor points of style that you should adhere to in your PR:
+
+- Use `=` for assignment (not `<-`); see [here](https://github.com/Rdatatable/data.table/pull/3582#discussion_r287075480), [here](https://github.com/Rdatatable/data.table/issues/3590#issuecomment-495776200), [here](https://stackoverflow.com/questions/1741820/what-are-the-differences-between-and-in-r#comment14293879_1742591) and [here](https://twitter.com/kdpsinghlab/status/1044568690346393606)
+- Spacing
+  + 2-space indentation
+  + No trailing white space
+  + In tests use long lines and vertical alignment to easily spot small differences
+  + Argument spacing style: `fun(arg1=value1, arg2=val2, ...)`
+  + Add a whitespace between `if` and opening bracket, also before opening curly bracket: `if (condition) {`
+- Use `L` suffix for integer; i.e. `x[1L]` not `x[1]` (to save coercion)
+- Use `stop(domain=NA, gettextf(fmt, arg1, arg2, ..., domain="R-data.table"))` not `stop(paste(...))` or `stop(sprintf(...))` to facilitate translation as per [Writing R Extensions#1.7 point 2](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Diagnostic-messages)
+
+### Translations
+
+`data.table` offers some translations for non-native English speakers (currently, it's limited to just Mandarin). For development, there's a few minor points:
+
+ - You _don't_ need to update the `po` files in your PR. We update these in one PR before each release. This makes it much easier for translators to update the messages in one go, rather than greenlighting each PR.
+ - For R code, when writing error messages (or `warning`s, or `message`s), use a single string where possible. Bad: `stop("This is ", "an error")` vs. Good: `stop("This is an error")`. If you need to construct the error message from several parts, use `gettextf` (and rarely, `ngettext`). Bad: `stop("Error in column ", j, ". Please fix."))` vs. Good: `stop(domain=NA, gettextf("Error in column %d. Please fix.", j, domain="R-data.table"))`. Note the usage of `domain` in both `stop()` and `gettextf()`. More uncommonly, when using `cat()`, always use `gettextf`.
+ - For C code, any `char` array that will be shown to users should be wrapped in `_()` (which marks the array for translation). Bad: `error("Error in column %d. Please fix.", j);` vs. Good: `error(_("Error in column %d. Please fix."), j);`.
```

Comment:
to be fair this sounds overly complicated, we can think if we could somehow improve that (not in this PR of course).